src/jemalloc.c

```diff
@@ -3420,12 +3407,13 @@ je_xallocx(void *ptr, size_t size, size_t extra, int flags) {
 	} else {
 		usize = ixallocx_helper(tsd_tsdn(tsd), ptr, old_usize, size,
 		    extra, alignment, zero);
+		thread_event(tsd, usize);
 	}
 	if (unlikely(usize == old_usize)) {
+		thread_event_rollback(tsd, usize);
```

Comment:
Yes. Rolling only once should be doable, but it can disturb the current code structure. See my Sep 25th comment.

Reply:
It's uncommon so should be fine. Let's add a test case with rolling back twice then?

Reply:
Do you have any particular suggestions on how to write the test case? It is pretty hard to write tests due to the reentrancy events: the tests need to assert things that can hold both with and without interrupting calls, and the tests don't have access to what happens in the interrupting calls, nor does it make sense for the tests to make any assumptions. So it's hard to assert on e.g. whether `thread_allocated` falls back to its original value. See my Sep 11th comments. Thoughts?

Reply:
I think explicitly call rollback twice in a row would be fine, to emulate the effect. Probably not worth the trouble setting up the exact scenario.

Reply:
Do you think the current test set in https://github.com/jemalloc/jemalloc/blob/dev/test/integration/xallocx.c is sufficient? The tests are basically examining all possible input scenarios for `xallocx()`. When `config_prof` and `config_debug` are on, when the `check_integration_prof` test set is running, the `thread_event_assert_invariants()` should be able to implicitly help us check about the counters?

Reply:
Let's still add some unit tests to explicitly do this. No need to be fancy, this may be fine:

    allocated = thread_allocated_get();
    rollback();
    some asserts;
    rollback();
    some asserts again;

Reply:
One difficulty is to determine what exactly we assert there (in your `some asserts` parts) - it needs to hold true either with or without the reentrancy setup. Earlier I figured that the only asserts I can add are something similar to those asserts I already have in `thread_event_assert_invariants()`, but those asserts are already tested if the build is in debug mode. Thoughts?

Reply:
Discussed offline. Will do low-level unit tests.